File: lang/cs/Org.Apache.REEF.Bridge.Core.Grpc/Client/ClientLauncher.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using System;
using System.IO;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Google.Protobuf;
using Grpc.Core;
using Org.Apache.REEF.Bridge.Core.Common.Client;
using Org.Apache.REEF.Bridge.Core.Common.Client.Config.Runtime.Proto;
using Org.Apache.REEF.Bridge.Core.Common.Driver;
using Org.Apache.REEF.Bridge.Core.Grpc.Driver;
using Org.Apache.REEF.Bridge.Core.Proto;
using Org.Apache.REEF.Common.Client;
using Org.Apache.REEF.Common.Files;
using Org.Apache.REEF.Tang.Annotations;
using Org.Apache.REEF.Tang.Formats;
using Org.Apache.REEF.Tang.Implementations.Configuration;
using Org.Apache.REEF.Tang.Interface;
using Org.Apache.REEF.Tang.Util;
using Org.Apache.REEF.Utilities;
using Org.Apache.REEF.Utilities.Logging;

namespace Org.Apache.REEF.Bridge.Core.Grpc.Client
{
    internal sealed class ClientLauncher : IClientLauncher
    {
        private const string SoFileNameExtension = ".so";
        private const string DylibFileNameExtension = ".dylib";
        private const string DllFileNameExtension = ".dll";
        private const string ExeFileNameExtension = ".exe";
        private const string JarFileNameExtension = ".jar";
        private const string ConfigFileNameExtension = ".config";

        private const string JavaClientLauncherClass = "org.apache.reef.bridge.client.grpc.ClientLauncher";

        private const string DriverExe = "Org.Apache.REEF.Bridge.Core.Driver.exe";
        private const string DriverDll = "Org.Apache.REEF.Bridge.Core.Driver.dll";

        private static readonly string jsonSDK = @"
{
  ""runtimeOptions"": {
        ""tfm"": ""netcoreapp2.0"",
        ""framework"": {
            ""name"": ""Microsoft.NETCore.App"",
            ""version"": ""2.0.0""
        }
    }
}
";

        private static readonly Logger Log = Logger.GetLogger(typeof(ClientLauncher));

        private readonly ClientService _clientService;

        private readonly DriverClientConfiguration _driverClientConfiguration;

        private readonly REEFFileNames _reefFileNames;

        private readonly IConfigurationSerializer _configurationSerializer;

        private readonly JavaClientLauncher _javaClientLauncher;

        private readonly Server _grpcServer;

        private readonly int _grpcServerPort;

        [Inject]
        private ClientLauncher(
            ClientService clientService,
            JavaClientLauncher javaClientLauncher,
            REEFFileNames reefFileNames,
            IConfigurationSerializer configurationSerializer,
            DriverClientParameters driverRuntimeProto,
            IRuntimeProtoProvider runtimeProtoProvider)
        {
            _clientService = clientService;
            _javaClientLauncher = javaClientLauncher;
            _reefFileNames = reefFileNames;
            _configurationSerializer = configurationSerializer;
            _driverClientConfiguration = driverRuntimeProto.Proto;
            runtimeProtoProvider.SetParameters(_driverClientConfiguration);
            _grpcServer = new Server
            {
                Services = { BridgeClient.BindService(clientService) },
                Ports = { new ServerPort("localhost", 0, ServerCredentials.Insecure) }
            };
            _grpcServer.Start();
            Log.Log(Level.Info, "Server port any {0}", _grpcServer.Ports.Any());
            foreach (var serverPort in _grpcServer.Ports)
            {
                Log.Log(Level.Info, "Server port {0}", serverPort.BoundPort);
                _grpcServerPort = serverPort.BoundPort;
            }
        }

        public void Dispose()
        {
            _grpcServer.ShutdownAsync();
        }

        public async Task<LauncherStatus> SubmitAsync(
            IConfiguration driverAppConfiguration, 
            CancellationToken cancellationToken)
        {
            var driverClientConfiguration =
                Configurations.Merge(driverAppConfiguration,
                    DriverBridgeConfiguration.ConfigurationModule
                        .Set(DriverBridgeConfiguration.DriverServiceClient, GenericType<DriverServiceClient>.Class)
                        .Set(DriverBridgeConfiguration.DriverClientService, GenericType<DriverClientService>.Class)
                        .Build());
            var jobFolder = Directory.CreateDirectory(Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString()));
            Log.Log(Level.Info, "Job folder {0}", jobFolder);
            Directory.CreateDirectory(jobFolder.FullName);
            AddDependencies(jobFolder, driverClientConfiguration);
            // Launch command
            if (File.Exists(DriverExe))
            {
                _driverClientConfiguration.DriverClientLaunchCommand = 
                    "cmd.exe /c " + Path.Combine(_reefFileNames.GetGlobalFolderPath(), DriverExe) + " " + 
                    _reefFileNames.GetClrDriverConfigurationPath();
            }
            else
            {
                _driverClientConfiguration.DriverClientLaunchCommand = 
                    "dotnet " + Path.Combine(_reefFileNames.GetGlobalFolderPath(), DriverDll) + " " + 
                    _reefFileNames.GetClrDriverConfigurationPath();
            }

            var driverClientConfigFile = Path.Combine(jobFolder.FullName, "driverclient.json");
            using (var outputFile = new StreamWriter(driverClientConfigFile))
            {
                outputFile.Write(JsonFormatter.Default.Format(_driverClientConfiguration));
            }
            // Submit a new job
            _clientService.Reset(); 
            var task = _javaClientLauncher.LaunchAsync(JavaLoggingSetting.Info,
                JavaClientLauncherClass,
                new[] { driverClientConfigFile, _grpcServerPort.ToString() },
                cancellationToken);
            lock (_clientService)
            {
                while (!_clientService.IsDone && !cancellationToken.IsCancellationRequested)
                {
                    Monitor.Wait(_clientService, TimeSpan.FromMinutes(1));
                }
            }
            await task;
            return _clientService.LauncherStatus;
        }

        private void AddDependencies(FileSystemInfo jobFolder, IConfiguration driverClientConfiguration)
        {
            // driver client configuration
            var driverClientConfigurationFile = Path.GetFullPath(Path.Combine(
                jobFolder.FullName,
                _reefFileNames.GetClrDriverConfigurationName()));
            _configurationSerializer.ToFile(driverClientConfiguration, driverClientConfigurationFile);
            _driverClientConfiguration.LocalFiles.Add(driverClientConfigurationFile);

            /*
            var jsonSDKFile = Path.Combine(jobFolder.FullName, "Org.Apache.REEF.Bridge.Core.Driver.runtimeconfig.json");
            File.WriteAllText(jsonSDKFile, jsonSDK);
            _driverClientConfiguration.GlobalFiles.Add(jsonSDKFile);
            */

            // resource files
            var a = typeof(ClientLauncher).Assembly;
            foreach (var name in a.GetManifestResourceNames())
            {
                Log.Log(Level.Info, "Extracting resource {0}", name);
                var resource = a.GetManifestResourceStream(name);
                using (var file = new FileStream(name, FileMode.Create, FileAccess.Write))
                {
                    resource.CopyTo(file);
                }
            }

            var directory = ClientUtilities.GetPathToExecutingAssembly();
            {
                if (Directory.Exists(directory))
                {
                    // For input paths that are directories, extract only files of a predetermined type
                    foreach (var assembly in Directory.GetFiles(directory).Where(IsAssemblyToCopy))
                    {
                        _driverClientConfiguration.GlobalFiles.Add(assembly);
                    }
                }
                else
                {
                    // Throw if a path input was not a file or a directory
                    throw new FileNotFoundException($"Global Assembly Directory not Found: {directory}");
                }
            }
        }

        /// <summary>
        /// Returns true, if the given file path references a DLL or EXE or JAR.
        /// </summary>
        /// <param name="filePath"></param>
        /// <returns></returns>
        private static bool IsAssemblyToCopy(string filePath)
        {
            var fileName = Path.GetFileName(filePath);
            if (string.IsNullOrWhiteSpace(fileName))
            {
                return false;
            }

            var lowerCasePath = fileName.ToLower();
            return lowerCasePath.EndsWith(DllFileNameExtension) ||
                   lowerCasePath.EndsWith(ExeFileNameExtension) ||
                   lowerCasePath.EndsWith(DylibFileNameExtension) ||
                   lowerCasePath.EndsWith(SoFileNameExtension) ||
                   lowerCasePath.EndsWith(ConfigFileNameExtension) ||
                   lowerCasePath.EndsWith(JarFileNameExtension);
        }
    }
}

File: lang/java/reef-bridge-proto-java/src/main/java/org/apache/reef/bridge/client/grpc/ClientLauncher.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.reef.bridge.client.grpc;

import com.google.protobuf.util.JsonFormat;
import org.apache.reef.bridge.driver.launch.RuntimeConfigurationProvider;
import org.apache.reef.bridge.driver.launch.azbatch.AzureBatchConfigurationProvider;
import org.apache.reef.bridge.driver.launch.hdinsight.HDInsightConfigurationProvider;
import org.apache.reef.bridge.driver.launch.local.LocalConfigurationProvider;
import org.apache.reef.bridge.driver.launch.yarn.YarnConfigurationProvider;
import org.apache.reef.bridge.driver.service.DriverServiceConfigurationProvider;
import org.apache.reef.bridge.driver.service.grpc.GRPCDriverServiceConfigurationProvider;
import org.apache.reef.bridge.proto.ClientProtocol;
import org.apache.reef.client.ClientConfiguration;
import org.apache.reef.client.LauncherStatus;
import org.apache.reef.tang.Configuration;
import org.apache.reef.tang.Configurations;
import org.apache.reef.tang.Tang;
import org.apache.reef.tang.exceptions.InjectionException;

import javax.inject.Inject;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Client launcher.
 */
public final class ClientLauncher {
  /**
   * Standard Java logger.
   */
  private static final Logger LOG = Logger.getLogger(ClientLauncher.class.getName());

  private static final Tang TANG = Tang.Factory.getTang();

  private final RuntimeConfigurationProvider runtimeConfigurationProvider;

  private final DriverServiceConfigurationProvider driverServiceConfigurationProvider;

  /**
   * This class should not be instantiated.
   */
  @Inject
  private ClientLauncher(
      final RuntimeConfigurationProvider runtimeConfigurationProvider,
      final DriverServiceConfigurationProvider driverServiceConfigurationProvider) {
    this.runtimeConfigurationProvider = runtimeConfigurationProvider;
    this.driverServiceConfigurationProvider = driverServiceConfigurationProvider;
  }

  private LauncherStatus launch(final int port,
      final ClientProtocol.DriverClientConfiguration driverClientConfigurationProto) {
    try {
      final Configuration runtimeConfiguration = Configurations.merge(
          getClientServiceConfiguration(port),
          this.runtimeConfigurationProvider.getRuntimeConfiguration(driverClientConfigurationProto));
      try (ClientService clientService = TANG.newInjector(runtimeConfiguration).getInstance(ClientService.class)) {
        return clientService.submit(
            this.driverServiceConfigurationProvider.getDriverServiceConfiguration(driverClientConfigurationProto));
      }
    } catch (final InjectionException ex) {
      LOG.log(Level.SEVERE, "Job configuration error", ex);
      throw new RuntimeException("Could not launch driver service", ex);
    }
  }

  static LauncherStatus submit(final int port,
      final ClientProtocol.DriverClientConfiguration driverClientConfiguration)
      throws InjectionException {
    final Configuration driverServiceLauncherConfiguration;
    switch (driverClientConfiguration.getRuntimeCase()) {
    case YARN_RUNTIME:
      driverServiceLauncherConfiguration = TANG.newConfigurationBuilder()
          .bindImplementation(DriverServiceConfigurationProvider.class,
              GRPCDriverServiceConfigurationProvider.class)
          .bind(RuntimeConfigurationProvider.class, YarnConfigurationProvider.class)
          .build();
      break;
    case LOCAL_RUNTIME:
      driverServiceLauncherConfiguration = TANG.newConfigurationBuilder()
          .bind(RuntimeConfigurationProvider.class, LocalConfigurationProvider.class)
          .bindImplementation(DriverServiceConfigurationProvider.class,
              GRPCDriverServiceConfigurationProvider.class)
          .build();
      break;
    case AZBATCH_RUNTIME:
      driverServiceLauncherConfiguration = TANG.newConfigurationBuilder()
          .bind(RuntimeConfigurationProvider.class, AzureBatchConfigurationProvider.class)
          .bindImplementation(DriverServiceConfigurationProvider.class,
              GRPCDriverServiceConfigurationProvider.class)
          .build();
      break;
    case HDI_RUNTIME:
      driverServiceLauncherConfiguration = TANG.newConfigurationBuilder()
          .bind(RuntimeConfigurationProvider.class, HDInsightConfigurationProvider.class)
          .bindImplementation(DriverServiceConfigurationProvider.class,
              GRPCDriverServiceConfigurationProvider.class)
          .build();
      break;
    default:
      throw new RuntimeException("unknown runtime " + driverClientConfiguration.getRuntimeCase());
    }
    return TANG.newInjector(driverServiceLauncherConfiguration).getInstance(ClientLauncher.class)
        .launch(port, driverClientConfiguration);
  }

  private static Configuration getClientServiceConfiguration(final int port) {
    return ClientServiceConfiguration.CONF
        .set(ClientServiceConfiguration.CLIENT_SERVER_PORT, port)
        .set(ClientConfiguration.ON_JOB_SUBMITTED, ClientService.SubmittedJobHandler.class)
        .set(ClientConfiguration.ON_JOB_RUNNING, ClientService.RunningJobHandler.class)
        .set(ClientConfiguration.ON_JOB_COMPLETED, ClientService.CompletedJobHandler.class)
        .set(ClientConfiguration.ON_JOB_FAILED, ClientService.FailedJobHandler.class)
        .set(ClientConfiguration.ON_JOB_MESSAGE, ClientService.JobMessageHandler.class)
        .set(ClientConfiguration.ON_RUNTIME_ERROR, ClientService.RuntimeErrorHandler.class)
        .set(ClientConfiguration.ON_WAKE_ERROR, ClientService.WakeErrorHandler.class)
        .build();
  }

  /**
   * Main method that launches the REEF job.
   *
   * @param args command line parameters.
   */
  public static void main(final String[] args) throws IOException, InjectionException {
    final ClientProtocol.DriverClientConfiguration driverClientConfigurationProto;
    if (args.length > 0) {
      final String content = new String(Files.readAllBytes(Paths.get(args[0])));
      final ClientProtocol.DriverClientConfiguration.Builder driverClientConfigurationProtoBuilder =
          ClientProtocol.DriverClientConfiguration.newBuilder();
      JsonFormat.parser()
          .usingTypeRegistry(JsonFormat.TypeRegistry.getEmptyTypeRegistry())
          .merge(content, driverClientConfigurationProtoBuilder);
      driverClientConfigurationProto = driverClientConfigurationProtoBuilder.build();
      final int port = (args.length == 2) ? Integer.parseInt(args[1]) : 0;
      final LauncherStatus status = submit(port, driverClientConfigurationProto);
      LOG.log(Level.INFO, "Status: {0}", status);
    } else {
      LOG.log(Level.SEVERE, ClientLauncher.class.getName() +
          " accepts single required argument referencing a file that contains a " +
          "client protocol buffer driver configuration");
      System.exit(1);
    }
  }
}
